src/language/ast.ts

```diff
@@ -360,6 +385,7 @@ export interface FieldNode {
   readonly arguments?: ReadonlyArray<ArgumentNode>;
   readonly directives?: ReadonlyArray<DirectiveNode>;
   readonly selectionSet?: SelectionSetNode;
+  readonly required: ComplexRequiredStatus;
```

Comment:
This should be a part of AST.
Without it you can't point to error.
For example, I wrote query like that:
```
{ 
   twoDimentionalArray[[[!]!]!]!
}
```
So error should be highlighted as follow:
```
{ 
   twoDimentionalArray[[[!]!]!]!
                        ^^^
}
```
If we go with "support array" root, then we should introduce 3 new AST nodes:
1. for `!`
2. for `?`
3. for `[]`

Also it will allow making property in `FieldNode` optional to represent "normal" state.
```
readonly requireModifiers?: RequireModifierNode;
```

Reply:
Also it will make `print` implementation simpler and make this feature compatible with other AST based function like `visit`.

Reply:
Thinking more about it, I'm not sure if how to design AST or if we should have 3 new nodes or one more complex node. 
But 100% sure it should be implemented as recursive AST nodes.
So please feel free to experiment with design and see what will fit our AST design.